File: python-webrtc/cpp/src/interfaces/rtc_audio_source.cpp
//
// Copyright 2022 Il`ya (Marshal) <https://github.com/MarshalX>. All rights reserved.
//
// Use of this source code is governed by a BSD-style license
// that can be found in the LICENSE.md file in the root of the project.
//

#include "rtc_audio_source.h"

#include <webrtc/rtc_base/ref_counted_object.h>

namespace python_webrtc {

  RTCAudioSource::RTCAudioSource() {
    _source = new rtc::RefCountedObject<RTCAudioTrackSource>();
  }

  void RTCAudioSource::Init(pybind11::module &m) {
    pybind11::class_<RTCAudioSource>(m, "RTCAudioSource")
        .def(pybind11::init<>())
        .def("createTrack", &RTCAudioSource::CreateTrack, pybind11::return_value_policy::reference)
        .def("onData", &RTCAudioSource::OnData);
  }

  MediaStreamTrack *RTCAudioSource::CreateTrack() {
    // TODO(mroberts): Again, we have some implicit factory we are threading around. How to handle?
    auto factory = PeerConnectionFactory::GetOrCreateDefault();
    auto track = factory->factory()->CreateAudioTrack(rtc::CreateRandomUuid(), _source.get());
    return MediaStreamTrack::holder()->GetOrCreate(factory, track);
  }

  void RTCAudioSource::OnData(RTCOnDataEvent &data) {
    _source->PushData(data);
  }

} // namespace python_webrtc

File: python-webrtc/python/webrtc/interfaces/rtc_audio_source.py
#
#  Copyright 2022 Il`ya (Marshal) <https://github.com/MarshalX>. All rights reserved.
#
#  Use of this source code is governed by a BSD-style license
#  that can be found in the LICENSE.md file in the root of the project.
#

from typing import TYPE_CHECKING

from webrtc import wrtc, WebRTCObject, MediaStreamTrack

if TYPE_CHECKING:
    import webrtc


class RTCAudioSource(WebRTCObject):
    """The :obj:`webrtc.MediaStreamTrack` interface represents a single media track within a stream;
    typically, these are audio or video tracks, but other track types may exist as well.
    """

    _class = wrtc.RTCAudioSource

    def create_track(self) -> 'webrtc.MediaStreamTrack':
        """Create :obj:`webrtc.MediaStreamTrack` whose source is the :obj:`webrtc.RTCAudioSource`.

        Returns:
            :obj:`webrtc.MediaStreamTrack`: A :obj:`webrtc.MediaStreamTrack` object representing the media track.
        """
        return MediaStreamTrack._wrap(self._native_obj.createTrack())

    def on_data(self, data: 'webrtc.RTCOnDataEvent'):
        """Push a new audio samples to every non-stopped local audio :obj:`webrtc.MediaStreamTrack`
        created with :attr:`createTrack`.

        Args:
            data (:obj:`webrtc.RTCOnDataEvent`): A :obj:`webrtc.RTCOnDataEvent` object representing new audio samples.

        Returns:
            :obj:`None`:
        """
        return self._native_obj.onData(data._native_obj)

    #: Alias for :attr:`create_track`
    createTrack = create_track
    #: Alias for :attr:`on_data`
    onData = on_data
